### PR TITLE
[Merged by Bors] - feat(NumberTheory/LSeries): even Hurwitz zeta functions (part 1)

### PR DESCRIPTION
**Pull request by loefflerd**
Add the construction of the even part of the Hurwitz zeta function (via FE pairs).

---

The un-completed version, and the relation to the defining Dirichlet series, will come in a later PR, in order to keep each PR in the series a manageable size.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean
Comment (CBirkbeck) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:61]: This is slightly shorter and quicker (although I can't figure out how to get the formatting of the preview to look right).

```suggestion
   have : cexp (-↑π * I * ((I * ↑x) + 2 * (↑ξ * I * ↑x))) = rexp (π * (x + 2 * ξ * x)) := by
        ring_nf
        simp only [I_sq, mul_neg, mul_one, neg_mul, neg_neg, sub_neg_eq_add, ofReal_exp, ofReal_add,
          ofReal_mul, ofReal_ofNat]
```

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean
Comment (CBirkbeck) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:149]:
```suggestion
  have h1 : I * ↑(1 / x) = -1 / (I * ↑x) := by
    push_cast; field_simp [hx.ne']; ring_nf; rw [I_sq]; ring_nf
  have h2 : (a' * I * x / (I * x)) = a' := by { field_simp [hx.ne']; ring_nf }
  have h3 : 1 / (-I * (I * x)) ^ (1 / 2 : ℂ) = 1 / ↑(x ^ (1 / 2 : ℝ)) := by
    rw [neg_mul, ← mul_assoc, I_mul_I, neg_one_mul, neg_neg,ofReal_cpow hx.le, ofReal_div, 
      ofReal_one, ofReal_ofNat]
  rw [h1, h2, h3]
  have : -↑π * I * (↑a' * I * ↑x) ^ 2 / (I * ↑x) = - (-π * a' ^ 2 * x) := by
    rw [mul_pow, mul_pow, I_sq]; field_simp [hx.ne']; ring_nf
  rw [this, ← mul_assoc, mul_comm (cexp _), mul_assoc _ (cexp _) (cexp _), ← Complex.exp_add,
      neg_add_self, Complex.exp_zero, mul_one, ofReal_div, ofReal_one]
```

**Review by CBirkbeck — COMMENTED** [review]
This is a far as I got before lunch, will come back later and look at the rest (but it all looks good!)

**Comment by CBirkbeck** [comment]
I just had a quick look at the rest and it all looks good to me :)

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:62]:
```suggestion
      congr
      ring).lift a
```
is a bit faster

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:139]:
```suggestion
  have hx' : I * x ≠ 0 := mul_ne_zero I_ne_zero (ofReal_ne_zero.mpr hx.ne')
  have h2 : (a * I * x / (I * x)) = a := by
    rw [div_eq_iff hx']
    ring
```

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:144]:
```suggestion
  have h4 : -π * I * (a * I * x) ^ 2 / (I * x) = - (-π * a ^ 2 * x) := by
    rw [mul_pow, mul_pow, I_sq, div_eq_iff hx']
    ring
```
Together with the previous one makes the proof twice as fast. (` field_simp`  is slow here.)

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:164]:
```suggestion
  have (n : ℤ) : ↑(rexp (-π * (↑n + a) ^ 2 * t)) =
      cexp (-↑π * ↑a ^ 2 * ↑t) * jacobiTheta₂_term n (↑a * I * ↑t) (I * ↑t) := by
    rw [jacobiTheta₂_term, ← Complex.exp_add]
    push_cast
    congr
    ring_nf
    simp only [I_sq, mul_neg, neg_mul, mul_one]
  simp only [this]
  exact (hasSum_jacobiTheta₂_term _ (by rwa [I_mul_im, ofReal_re])).mul_left _
```
Slow ` convert` .

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:173]:
```suggestion
  have (n : ℤ) : cexp (2 * ↑π * I * ↑a * ↑n) * ↑(rexp (-π * ↑n ^ 2 * t)) =
      jacobiTheta₂_term n (↑a) (I * ↑t) := by
    rw [jacobiTheta₂_term, ofReal_exp, ← Complex.exp_add]
    push_cast
    ring_nf
    simp only [I_sq, mul_neg, neg_mul, mul_one, sub_eq_add_neg]
  simp only [this]
  exact hasSum_jacobiTheta₂_term a (by rwa [I_mul_im, ofReal_re] : 0 < im (I * t))
```
Dito.

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:183]:
```suggestion
    simpa only [Int.cast_add, neg_mul, Int.cast_neg, add_left_neg, ne_eq, OfNat.ofNat_ne_zero,
      not_false_eq_true, zero_pow, mul_zero, zero_mul, Real.exp_zero]
      using hasSum_ite_sub_hasSum (hasSum_int_evenKernel (k : ℝ) ht) (-k)
```
is faster, but longer...

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:191]:
```suggestion
    (↑(cosKernel a t) - 1) := by
  simpa using hasSum_ite_sub_hasSum (hasSum_int_cosKernel a ht) 0
```
is easier to read (one could also squeeze the ` simpa` ...).

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:203]: This ` convert`  is also slow, but the resulting equation is maybe too long to state explicitly...

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:206]:
```suggestion
  congr 3 <;> ring
```

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:225]: Is there a reason to call the positive real whose existence is asserted ` r`  above and ` p`  here?

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:270]:
```suggestion
    have (x : ℝ) : ‖(ofReal' ∘ cosKernel a) x - 1‖ = ‖cosKernel a x - 1‖ := by
      rw [← norm_real, ofReal_sub, ofReal_one, Function.comp_apply]
    simp only [this]
    exact hp'.trans (isLittleO_exp_neg_mul_rpow_atTop hp _).isBigO
```
as usual...

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:256]:
```suggestion
  hk := one_half_pos
```

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:330]:
```suggestion
  · rw [completedCosZeta₀, WeakFEPair.symm, hurwitzEvenFEPair, smul_eq_mul, mul_one, div_div,
```

**Review by MichaelStollBayreuth — APPROVED** [review]
Apart from a few performance issues, this looks good!

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:218]: Now you can use the new congruence lemma here.

Comment (loefflerd) [reply Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean]: Of course. I was just waiting for CI to build all the files in between (I should probably have applied the awaiting-CI tag.) Done now.

**Comment by MichaelStollBayreuth** [comment]
maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**
